Add move and cancel by ID for PI modification to APIs

tmetzke opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #41874, #39819

tmetzke committed:
feat: add PI modify move and cancel by ID to API protocol

Adds the move and cancel by ID options for process instance modification to the API protocols (gRPC and REST V2).

tmetzke committed:
feat: add PI modify move and cancel by ID to gRPC

Adds the move and cancel by ID options for process instance modification to the gRPC API.

tmetzke committed:
test: add PI modify move and cancel by ID to gRPC

Ensure the move and cancel by ID options for process instance modification work as expected in the gRPC API.

tmetzke committed:
feat: add PI modify move and cancel by ID to REST

Adds the move and cancel by ID options for process instance modification to the REST API.
Refactors requests with alternatives (using oneOf), introducing a common abstract class and exception handling for them.

tmetzke committed:
test: refactor REST requests with alternatives

Ensure requests with alternatives (using oneOf) work as expected. The exception responses are now back to 8.8-level.
This ensures https://github.com/camunda/camunda/issues/39819 is fixed.

tmetzke committed:
feat: add move instructions to service layer

tmetzke committed:
feat: add PI modify move and cancel by ID to Java SDK

Adds the move and cancel by ID options for process instance modification to the Java SDK.

tmetzke committed:
feat: adjust Operate V1 to new service layer

tmetzke committed:
fix: keep batch modification API independent

The batch process instance modification API must stay backward-compatible.
Thus, it needs to be decoupled from the new move instructions. They require more attributes than the batch API does.

tmetzke committed:
refactor: remove superficial semi-colon

nicpuppa reviewed (commented):
@tmetzke looks good from specification and client side

tmetzke committed:
docs: improve PI V2 API descriptions

tmetzke committed:
feat: make ancestor scope instruction optional for move

Removes ancestor scope instructions from required properties for move instructions.
The default is using a "direct" scope with key "-1".

Provides a simpler message for mismatched type deserialization in Jackson based on defined discriminator values.

tmetzke committed:
test: validate PI modification in REST API

Ensures process instance modification requests adhere to the specification.
Extends the happy path instructions to cover all scenarios.
Extends rejection tests for terminate instructions.
Introduces rejection tests for move instructions.

tmetzke committed:
feat: support variables for move instruction in Java SDK

tmetzke committed:
test: validate PI modification in Java SDK

Ensures process instance modification requests contain expected data when constructed in the Java SDK.

tmetzke committed:
test: validate PI modification E2E

Ensures process instance modification requests work end-to-end with an integration test for move.

tmetzke commented on clients/java/src/test/java/io/camunda/client/process/rest/ModifyProcessInstanceRestTest.java:
```diff
@@ -108,66 +111,6 @@ public void shouldActivateMultipleElementsWithVariables() {
-        request.getTerminateInstructions().get(0);
-    assertTerminateInstruction(terminateInstruction, ELEMENT_KEY_A);
-  }
-
```
ℹ️ I simply moved those further down to better group the tests by activate, terminate, move, and mixed commands.

tmetzke commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/GlobalControllerExceptionHandler.java:
```diff
@@ -105,6 +127,12 @@ private boolean isRequestBodyMissing(final Exception ex) {
+  private boolean isArrayTypeDeserializationError(final Exception ex) {
+    return ex instanceof final HttpMessageNotReadableException exception
+        && exception.getRootCause() != null
+        && exception.getRootCause() instanceof final DeserializationException dEx;
```
ℹ️ We throw these from our custom deserializers and should pick them up here

tmetzke committed:
docs: align property description

tmetzke committed:
feat: add variable validation for move modification

tmetzke committed:
docs: fix typo in PI spec

tsedekey reviewed (approved):
I see you did a few clean-ups along the way which is 😎 Overall LGTM 🚀
